File: app/src/main/java/com/robifr/ledger/ui/product/viewmodel/ProductFilterViewModel.java
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.product.viewmodel;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.util.Pair;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import com.robifr.ledger.data.display.ProductFilterer;
import com.robifr.ledger.data.display.ProductFilters;
import com.robifr.ledger.data.model.ProductModel;
import com.robifr.ledger.util.CurrencyFormat;
import com.robifr.ledger.util.livedata.SafeLiveData;
import com.robifr.ledger.util.livedata.SafeMutableLiveData;
import java.math.BigDecimal;
import java.text.ParseException;
import java.util.List;
import java.util.Objects;

public class ProductFilterViewModel {
  @NonNull private final ProductViewModel _viewModel;
  @NonNull private final ProductFilterer _filterer;

  @NonNull
  private final SafeMutableLiveData<String> _inputtedMinPriceText = new SafeMutableLiveData<>("");

  @NonNull
  private final SafeMutableLiveData<String> _inputtedMaxPriceText = new SafeMutableLiveData<>("");

  public ProductFilterViewModel(
      @NonNull ProductViewModel viewModel, @NonNull ProductFilterer filterer) {
    this._viewModel = Objects.requireNonNull(viewModel);
    this._filterer = Objects.requireNonNull(filterer);
  }

  @NonNull
  public SafeLiveData<String> inputtedMinPriceText() {
    return this._inputtedMinPriceText;
  }

  @NonNull
  public SafeLiveData<String> inputtedMaxPriceText() {
    return this._inputtedMaxPriceText;
  }

  /**
   * Get current inputted filter from any corresponding inputted live data. If any live data is set
   * using {@link MutableLiveData#postValue(Object)}, calling this method may not immediately
   * reflect the latest changes. For accurate results in asynchronous operations, consider calling
   * this method inside {@link Observer}.
   */
  @NonNull
  public ProductFilters inputtedFilters() {
    // Nullable value to represent unbounded range.
    Long minPrice = null;
    Long maxPrice = null;

    try {
      if (!this._inputtedMinPriceText.getValue().isBlank()) {
        minPrice =
            CurrencyFormat.parse(
                    this._inputtedMinPriceText.getValue(),
                    AppCompatDelegate.getApplicationLocales().toLanguageTags())
                .longValue();
      }

    } catch (ParseException ignore) {
    }

    try {
      if (!this._inputtedMaxPriceText.getValue().isBlank()) {
        maxPrice =
            CurrencyFormat.parse(
                    this._inputtedMaxPriceText.getValue(),
                    AppCompatDelegate.getApplicationLocales().toLanguageTags())
                .longValue();
      }

    } catch (ParseException ignore) {
    }

    return ProductFilters.toBuilder().setFilteredPrice(new Pair<>(minPrice, maxPrice)).build();
  }

  public void onMinPriceTextChanged(@NonNull String minPrice) {
    Objects.requireNonNull(minPrice);

    this._inputtedMinPriceText.setValue(minPrice);
  }

  public void onMaxPriceTextChanged(@NonNull String maxPrice) {
    Objects.requireNonNull(maxPrice);

    this._inputtedMaxPriceText.setValue(maxPrice);
  }

  public void onFiltersChanged(@NonNull ProductFilters filters) {
    this.onFiltersChanged(filters, this._viewModel.products().getValue());
  }

  public void onFiltersChanged(
      @NonNull ProductFilters filters, @NonNull List<ProductModel> products) {
    Objects.requireNonNull(filters);
    Objects.requireNonNull(products);

    final String minTotalPrice =
        filters.filteredPrice().first != null
            ? CurrencyFormat.format(
                BigDecimal.valueOf(filters.filteredPrice().first),
                AppCompatDelegate.getApplicationLocales().toLanguageTags())
            : "";
    final String maxTotalPrice =
        filters.filteredPrice().second != null
            ? CurrencyFormat.format(
                BigDecimal.valueOf(filters.filteredPrice().second),
                AppCompatDelegate.getApplicationLocales().toLanguageTags())
            : "";

    this.onMinPriceTextChanged(minTotalPrice);
    this.onMaxPriceTextChanged(maxTotalPrice);
    this._filterer.setFilters(filters);

    final List<ProductModel> filteredProducts = this._filterer.filter(products);
    // Re-sort the list, after previously re-populating the list with a new filtered value.
    this._viewModel.onSortMethodChanged(this._viewModel.sortMethod().getValue(), filteredProducts);
  }
}

File: app/src/main/java/com/robifr/ledger/ui/product/viewmodel/ProductFilterViewModel.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.product.viewmodel

import androidx.appcompat.app.AppCompatDelegate
import com.robifr.ledger.data.display.ProductFilters
import com.robifr.ledger.ui.common.state.SafeLiveData
import com.robifr.ledger.ui.common.state.SafeMutableLiveData
import com.robifr.ledger.util.CurrencyFormat
import java.text.ParseException

class ProductFilterViewModel(private val _viewModel: ProductViewModel) {
  private val _uiState: SafeMutableLiveData<ProductFilterState> =
      SafeMutableLiveData(
          ProductFilterState(isDialogShown = false, formattedMinPrice = "", formattedMaxPrice = ""))
  val uiState: SafeLiveData<ProductFilterState>
    get() = _uiState

  fun onMinPriceTextChanged(formattedMinPrice: String) {
    _uiState.setValue(_uiState.safeValue.copy(formattedMinPrice = formattedMinPrice))
  }

  fun onMaxPriceTextChanged(formattedMaxPrice: String) {
    _uiState.setValue(_uiState.safeValue.copy(formattedMaxPrice = formattedMaxPrice))
  }

  fun onDialogShown() {
    _uiState.setValue(_uiState.safeValue.copy(isDialogShown = true))
  }

  fun onDialogClosed() {
    _uiState.setValue(_uiState.safeValue.copy(isDialogShown = false))
    _onFiltersChanged()
    _viewModel.onReloadPage(1, 1)
  }

  fun parseInputtedFilters(): ProductFilters {
    // All these nullable value to represent unbounded range.
    var minPrice: Long? = null
    try {
      if (_uiState.safeValue.formattedMinPrice.isNotBlank()) {
        minPrice =
            CurrencyFormat.parseToCents(
                    _uiState.safeValue.formattedMinPrice,
                    AppCompatDelegate.getApplicationLocales().toLanguageTags())
                .toLong()
      }
    } catch (_: ParseException) {}

    var maxPrice: Long? = null
    try {
      if (_uiState.safeValue.formattedMaxPrice.isNotBlank()) {
        maxPrice =
            CurrencyFormat.parseToCents(
                    _uiState.safeValue.formattedMaxPrice,
                    AppCompatDelegate.getApplicationLocales().toLanguageTags())
                .toLong()
      }
    } catch (_: ParseException) {}

    return ProductFilters(minPrice to maxPrice)
  }

  private fun _onFiltersChanged(filters: ProductFilters = parseInputtedFilters()) {
    onMinPriceTextChanged(
        filters.filteredPrice.first?.let {
          CurrencyFormat.formatCents(
              it.toBigDecimal(), AppCompatDelegate.getApplicationLocales().toLanguageTags())
        } ?: "")
    onMaxPriceTextChanged(
        filters.filteredPrice.second?.let {
          CurrencyFormat.formatCents(
              it.toBigDecimal(), AppCompatDelegate.getApplicationLocales().toLanguageTags())
        } ?: "")
  }
}
